src/neural-network.js

```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+    if (this.sizes === null) {
+      let sizes = this._getSizesFromData (data);
+      this.initialize (sizes);
     }
 
-    let error = 1;
-    let i;
-    for (i = 0; i < iterations && error > errorThresh; i++) {
+    while (res.iterations < iterations && res.error > errorThresh && Date.now () > endTime) {
+      res.iterations++;
       let sum = 0;
-      for (let j = 0; j < data.length; j++) {
-        let err = this.trainPattern(data[j].input, data[j].output, learningRate);
-        sum += err;
+      for (var i = 0; i < data.length; ++i) {
+        sum += this.trainPattern (data[i].input, data[i].output, learningRate);
       }
-      error = sum / data.length;
 
-      if (log && (i % logPeriod === 0)) {
-        log('iterations:', i, 'training error:', error);
+      res.error = sum / data.length;
+
+      if (log && (res.iterations % logPeriod === 0)) {
+        log('iterations:', res.iterations, 'training error:', res.error);
       }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
+        // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+        callback (JSON.parse (JSON.stringify (res)));
       }
     }
+    return res;
+  }
 
-    return {
-      error: error,
-      iterations: i
+  /**
+   *
+   * @param data
+   * @param _options
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
+    if (typeof _options === "function") {
+      cb = _options;
+      _options = {};
+    }
+    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+    data = this.formatData(data);
+    let iterations = options.iterations;
+    let errorThresh = options.errorThresh;
+    let log = options.log === true ? console.log : options.log;
+    let logPeriod = options.logPeriod;
+    let learningRate = _options.learningRate || this.learningRate || options.learningRate;
+    let callback = options.callback;
+    let callbackPeriod = options.callbackPeriod;
+    let endTime = Date.now() + options.trainTimeMs;
+    let res = {
+      error: 1,
+      iterations: 0
     };
+
+    if (this.sizes === null) {
+      let sizes = this._getSizesFromData (data);
+      this.initialize (sizes);
+    }
+
+    const items = new Array(iterations);
+    const thaw  = new Thaw (items, {
+      delay: true,
+      each: () => {
+        res.iterations++;
+        let sum = 0;
+        for (var i = 0; i < data.length; ++i) {
+          sum += this.trainPattern (data[i].input, data[i].output, learningRate);
+        }
+
+        res.error = sum / data.length;
+
+        if (log && (res.iterations % logPeriod === 0)) {
+          log(`iterations: ${res.iterations} training error: ${res.error}`);
+        }
+
+        if (callback && (res.iterations % callbackPeriod === 0)) {
```

Comment:
Like above, the callback is for partially trained nets?

Reply:
that is correct as well